packages/expo-modules-core/android/src/main/java/expo/modules/networkfetch/NativeResponse.kt

```diff
@@ -0,0 +1,225 @@
+    private set(value) {
+      synchronized(this) {
+        field = value
+      }
+      coroutineScope.launch {
+        this@NativeResponse.stateChangeOnceListeners.removeAll { it(value) }
+      }
+    }
+  private val stateChangeOnceListeners: MutableList<StateChangeListener> = mutableListOf()
+
+  var responseInit: NativeResponseInit? = null
+    private set
+  var error: Exception? = null
+    private set
+
+  val bodyUsed: Boolean
+    get() = this.ref.bodyUsed
+
+  override fun deallocate() {
+    this.ref.finalize()
+    super.deallocate()
+  }
+
+  fun startStreaming() {
+    if (isInvalidState(listOf(ResponseState.RESPONSE_RECEIVED, ResponseState.BODY_COMPLETED))) {
+      return
+    }
+    if (state == ResponseState.RESPONSE_RECEIVED) {
+      state = ResponseState.BODY_STREAMING_STARTED
+      val queuedData = this.ref.finalize()
+      emit("didReceiveResponseData", queuedData)
+    } else if (state == ResponseState.BODY_COMPLETED) {
+      val queuedData = this.ref.finalize()
+      emit("didReceiveResponseData", queuedData)
+      emit("didComplete")
+    }
+  }
+
+  fun cancelStreaming() {
+    if (isInvalidState(listOf(ResponseState.BODY_STREAMING_STARTED))) {
+      return
+    }
+    state = ResponseState.BODY_STREAMING_CANCELLED
+  }
+
+  fun emitRequestCancelled() {
+    error = NetworkFetchRequestCancelledException()
+    state = ResponseState.ERROR_RECEIVED
+  }
+
+  fun waitForStates(states: List<ResponseState>, callback: (ResponseState) -> Unit) {
+    if (states.contains(state)) {
+      callback(state)
+      return
+    }
+    stateChangeOnceListeners.add { newState ->
+      if (states.contains(newState)) {
+        callback(newState)
+        return@add true
+      }
+      return@add false
+    }
+  }
+
+  //region Callback implementations
+
+  override fun onFailure(call: Call, e: IOException) {
+    if (isInvalidState(
+        listOf(
+          ResponseState.STARTED,
+          ResponseState.RESPONSE_RECEIVED,
+          ResponseState.BODY_STREAMING_STARTED,
+          ResponseState.BODY_STREAMING_CANCELLED
+        )
+      )
+    ) {
+      return
+    }
+
+    if (state == ResponseState.BODY_STREAMING_STARTED) {
+      emit("didFailWithError", e)
+    }
+    error = e
+    state = ResponseState.ERROR_RECEIVED
+  }
+
+  override fun onResponse(call: Call, response: Response) {
+    responseInit = createResponseInit(response)
+    state = ResponseState.RESPONSE_RECEIVED
+
+    coroutineScope.launch(Dispatchers.IO) {
+      val stream = response.body?.source() ?: return@launch
+      pumpResponseBodyStream(stream)
+      response.close()
+
+      if (this@NativeResponse.state == ResponseState.BODY_STREAMING_STARTED) {
+        emit("didComplete")
+      }
+      this@NativeResponse.state = ResponseState.BODY_COMPLETED
+    }
+  }
+
+  //endregion Callback implementations
+
+  //region Internals
+
+  private fun isInvalidState(validStates: List<ResponseState>): Boolean {
```

Comment:
```suggestion
  private fun isInvalidState(varargs validStates: ResponseState): Boolean {
```
It's your choice, but it might be easier to call that function.